walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,67 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.ImmutableList; 
 
+/** Finds list of all potential calendar time slots to schedule a meeting with the given 
+ * MeetingRequest constraints. Accounts for optional attendees when possible. */
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ImmutableList<TimeRange> conflicting = getRelevantEvents(request.getAttendees(), events);
+    ImmutableList<TimeRange> times = getMeetingTimes(request.getDuration(), conflicting);
+    
+    Iterable<String> allAttendeesIter = 
+        Iterables.unmodifiableIterable(Iterables.concat(request.getAttendees(),
+                                                        request.getOptionalAttendees()));
+    Collection<String> allAttendees = Lists.newArrayList(allAttendeesIter);
+    
+    ImmutableList<TimeRange> optionalConflicting = getRelevantEvents(allAttendees, events);
+    ImmutableList<TimeRange> timesWithOptional = getMeetingTimes(request.getDuration(), optionalConflicting);
+
+    if (timesWithOptional.size() == 0 && request.getAttendees().size() != 0) {
+      return times;
+    } else {
+      return timesWithOptional;
+    }
+  }
+
+  // Returns list of events that at least one of given attendees is going to.
+  private ImmutableList<TimeRange> getRelevantEvents(Collection<String> attendees, Collection<Event> events) {
+    List<TimeRange> conflicting = new ArrayList<TimeRange>();
+    for (Event evt : events) {
+       if (evt.getAttendees().stream().anyMatch(attendees::contains)) {
+        conflicting.add(evt.getWhen());
+      }
+    }
+    Collections.sort(conflicting, TimeRange.ORDER_BY_START);
+    return ImmutableList.copyOf(conflicting);
+  }
+
+  // Returns list of time slots that fit the given list of potentially conflicting events.
+  private ImmutableList<TimeRange> getMeetingTimes(long duration, List<TimeRange> conflicting) {
+    List<TimeRange> times = new ArrayList<TimeRange>();
+    int busyUntil = TimeRange.START_OF_DAY;
+    for (TimeRange time : conflicting) {
```

Comment:
optional: if you're interested, maybe you can try to see if you could rewrite this via the Stream notation. Not necessary though, especially since this looks really good and is easily understandable.